File: lib/binaries.ts
import fs from 'fs';
import path from 'path';
import https from 'https';
import { execFileSync } from 'child_process';
import { getSetting, getDataDir } from './db';

export type BinaryName = 'ytdlp' | 'gallery-dl' | 'ffmpeg';

// ── Paths ────────────────────────────────────────────────────────────────────

export function getBinDir(): string {
  return path.join(getDataDir(), 'bin');
}

function defaultBinaryPath(name: BinaryName): string {
  const dir = getBinDir();
  if (name === 'ytdlp') return path.join(dir, ytdlpFilename());
  if (name === 'gallery-dl') return path.join(dir, galleryDlFilename());
  return path.join(dir, ffmpegFilename());
}

export function getYtdlpPath(): string {
  const override = getSetting('ytdlp_bin');
  return override?.trim() || defaultBinaryPath('ytdlp');
}

export function getGalleryDlPath(): string {
  const override = getSetting('gallerydl_bin');
  return override?.trim() || defaultBinaryPath('gallery-dl');
}

export function getFfmpegPath(): string {
  const override = getSetting('ffmpeg_bin');
  return override?.trim() || defaultBinaryPath('ffmpeg');
}

export function getFfprobePath(): string {
  const ffprobeName = process.platform === 'win32' ? 'ffprobe.exe' : 'ffprobe';
  return path.join(getBinDir(), ffprobeName);
}

// ── Platform helpers ─────────────────────────────────────────────────────────

function ytdlpFilename(): string {
  if (process.platform === 'win32') return 'yt-dlp.exe';
  if (process.platform === 'darwin') return 'yt-dlp_macos';
  if (process.arch === 'arm64' || process.arch === 'arm') return 'yt-dlp_linux_aarch64';
  return 'yt-dlp';
}

function galleryDlFilename(): string {
  if (process.platform === 'win32') return 'gallery-dl.exe';
  if (process.platform === 'darwin') return 'gallery-dl';
  return 'gallery-dl.bin'; // Linux
}

function ffmpegFilename(): string {
  return process.platform === 'win32' ? 'ffmpeg.exe' : 'ffmpeg';
}

function ytdlpDownloadUrl(): string {
  return `https://github.com/yt-dlp/yt-dlp/releases/latest/download/${ytdlpFilename()}`;
}

function galleryDlDownloadUrl(): string {
  return `https://github.com/mikf/gallery-dl/releases/latest/download/${galleryDlFilename()}`;
}

function ffmpegArchiveUrl(): string {
  if (process.platform === 'darwin') {
    // evermeet.cx provides a zip containing just the ffmpeg binary
    return 'https://evermeet.cx/ffmpeg/getrelease/ffmpeg/zip';
  }
  if (process.platform === 'win32') {
    return 'https://github.com/yt-dlp/FFmpeg-Builds/releases/latest/download/ffmpeg-master-latest-win64-gpl.zip';
  }
  // Linux
  if (process.arch === 'arm64' || process.arch === 'arm') {
    return 'https://github.com/yt-dlp/FFmpeg-Builds/releases/latest/download/ffmpeg-master-latest-linuxarm64-gpl.tar.xz';
  }
  return 'https://github.com/yt-dlp/FFmpeg-Builds/releases/latest/download/ffmpeg-master-latest-linux64-gpl.tar.xz';
}

function ffmpegArchiveExt(): string {
  return process.platform === 'linux' ? '.tar.xz' : '.zip';
}

// ── Status ───────────────────────────────────────────────────────────────────

export interface BinaryStatus {
  name: BinaryName;
  path: string;
  exists: boolean;
  version: string | null;
}

export function checkBinary(name: BinaryName): BinaryStatus {
  const binPath =
    name === 'ytdlp' ? getYtdlpPath() :
    name === 'gallery-dl' ? getGalleryDlPath() :
    getFfmpegPath();

  const exists = fs.existsSync(binPath);
  let version: string | null = null;

  if (exists) {
    try {
      // ffmpeg uses -version (single dash), others use --version
      const versionFlag = name === 'ffmpeg' ? '-version' : '--version';
      const out = execFileSync(binPath, [versionFlag], {
        timeout: 5000,
        encoding: 'utf-8',
        stdio: ['ignore', 'pipe', 'pipe'],
      }).trim();
      // ffmpeg: "ffmpeg version 7.1 Copyright..."  — take first token after "version "
      // others: first line is just the version string
      if (name === 'ffmpeg') {
        const m = out.match(/version\s+(\S+)/);
        version = m ? m[1] : out.split('\n')[0].trim();
      } else {
        version = out.split('\n')[0].trim();
      }
    } catch (err) {
      console.warn(`[memevaultproject] checkBinary(${name}): version check failed:`, err instanceof Error ? err.message : err);
    }
  }

  return { name, path: binPath, exists, version };
}

// ── Download ─────────────────────────────────────────────────────────────────

export async function downloadBinary(name: BinaryName): Promise<BinaryStatus> {
  if (name === 'ffmpeg') return downloadFfmpeg();

  const url = name === 'ytdlp' ? ytdlpDownloadUrl() : galleryDlDownloadUrl();
  const binDir = getBinDir();
  fs.mkdirSync(binDir, { recursive: true });

  const destPath = defaultBinaryPath(name);
  const tmpPath = destPath + '.tmp';

  await downloadFile(url, tmpPath);

  if (process.platform !== 'win32') {
    fs.chmodSync(tmpPath, 0o755);
  }

  fs.renameSync(tmpPath, destPath);
  return checkBinary(name);
}

async function downloadFfmpeg(): Promise<BinaryStatus> {
  const binDir = getBinDir();
  fs.mkdirSync(binDir, { recursive: true });

  const archiveExt = ffmpegArchiveExt();
  const tmpArchive = path.join(binDir, `ffmpeg_download${archiveExt}`);
  const tmpExtractDir = path.join(binDir, 'ffmpeg_extract_tmp');
  const destPath = defaultBinaryPath('ffmpeg');

  try {
    // 1. Download the archive
    await downloadFile(ffmpegArchiveUrl(), tmpArchive);

    // 2. Extract to a temp directory
    fs.mkdirSync(tmpExtractDir, { recursive: true });
    extractArchive(tmpArchive, tmpExtractDir);

    // 3. Find the ffmpeg binary inside the extracted tree
    const found = findFile(tmpExtractDir, ffmpegFilename());
    if (!found) {
      throw new Error(`Could not find ${ffmpegFilename()} in the downloaded archive`);
    }

    // 4. Move to final destination and make executable
    fs.copyFileSync(found, destPath);
    if (process.platform !== 'win32') {
      fs.chmodSync(destPath, 0o755);
    }

    // 5. Also extract ffprobe if present in the archive
    const ffprobeName = process.platform === 'win32' ? 'ffprobe.exe' : 'ffprobe';
    const ffprobeDestPath = path.join(binDir, ffprobeName);
    const foundFfprobe = findFile(tmpExtractDir, ffprobeName);
    if (foundFfprobe) {
      fs.copyFileSync(foundFfprobe, ffprobeDestPath);
      if (process.platform !== 'win32') {
        fs.chmodSync(ffprobeDestPath, 0o755);
      }
    }
  } finally {
    // Clean up temp files regardless of success/failure
    try { fs.rmSync(tmpArchive, { force: true }); } catch { /* ignore */ }
    try { fs.rmSync(tmpExtractDir, { recursive: true, force: true }); } catch { /* ignore */ }
  }

  return checkBinary('ffmpeg');
}

function extractArchive(archivePath: string, destDir: string): void {
  const ext = archivePath.endsWith('.tar.xz') ? '.tar.xz' : '.zip';

  if (ext === '.tar.xz') {
    execFileSync('tar', ['-xf', archivePath, '-C', destDir]);
  } else if (process.platform === 'win32') {
    execFileSync('powershell', [
      '-Command',
      `Expand-Archive -Path '${archivePath}' -DestinationPath '${destDir}' -Force`,
    ]);
  } else {
    // macOS / Linux with zip
    execFileSync('unzip', ['-o', archivePath, '-d', destDir]);
  }
}

/** Recursively find the first file matching `filename` inside `dir`. */
function findFile(dir: string, filename: string): string | null {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      const found = findFile(full, filename);
      if (found) return found;
    } else if (entry.name === filename) {
      return full;
    }
  }
  return null;
}

/** Follow redirects and download a URL to a file path. */
function downloadFile(url: string, dest: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const out = fs.createWriteStream(dest);

    function get(currentUrl: string, redirects = 0) {
      if (redirects > 10) return reject(new Error('Too many redirects'));

      https
        .get(currentUrl, (res) => {
          if (
            res.statusCode &&
            res.statusCode >= 300 &&
            res.statusCode < 400 &&
            res.headers.location
          ) {
            get(res.headers.location, redirects + 1);
            return;
          }

          if (res.statusCode !== 200) {
            out.destroy();
            fs.unlink(dest, () => {});
            return reject(new Error(`HTTP ${res.statusCode} downloading ${currentUrl}`));
          }

          res.pipe(out);
          out.on('finish', () => out.close(() => resolve()));
          out.on('error', (err) => {
            out.destroy();
            fs.unlink(dest, () => {});
            reject(err);
          });
        })
        .on('error', (err) => {
          out.destroy();
          fs.unlink(dest, () => {});
          reject(err);
        });
    }

    get(url);
  });
}

// ── Startup helper ───────────────────────────────────────────────────────────

/**
 * Called at server startup. Auto-downloads any managed binary that is missing
 * (i.e. no custom path override is set and the default path doesn't exist).
 * Runs in the background — never throws.
 */
export async function ensureBinaries(): Promise<void> {
  const entries: { name: BinaryName; settingKey: string }[] = [
    { name: 'ytdlp',       settingKey: 'ytdlp_bin' },
    { name: 'gallery-dl',  settingKey: 'gallerydl_bin' },
    { name: 'ffmpeg',      settingKey: 'ffmpeg_bin' },
  ];

  for (const { name, settingKey } of entries) {
    const hasOverride = !!getSetting(settingKey)?.trim();
    if (hasOverride) continue;

    const status = checkBinary(name);
    if (!status.exists) {
      console.log(`[memevaultproject] ${name} not found, downloading…`);
      try {
        const result = await downloadBinary(name);
        console.log(`[memevaultproject] ${name} downloaded: ${result.path} (${result.version})`);
      } catch (err) {
        console.error(`[memevaultproject] Failed to download ${name}:`, err);
      }
    }
  }
}

File: src/binaries.rs
//! Auto-managed yt-dlp, gallery-dl, ffmpeg/ffprobe binaries.

use crate::config::Config;
use crate::db::{self, Db};
use serde::Serialize;
use std::path::Path;
use std::process::Stdio;

#[derive(Debug, Clone, Serialize)]
pub struct BinaryStatus {
    pub name: String,
    pub path: String,
    pub exists: bool,
    pub version: Option<String>,
}

pub fn get_ytdlp_path(db: &Db, config: &Config) -> String {
    let override_path = db
        .with_conn(|c| Ok(db::get_setting(c, "ytdlp_bin")))
        .ok()
        .flatten()
        .filter(|s| !s.is_empty());
    override_path.unwrap_or_else(|| {
        config
            .bin_dir()
            .join(ytdlp_filename())
            .to_string_lossy()
            .into()
    })
}

pub fn get_gallerydl_path(db: &Db, config: &Config) -> String {
    let override_path = db
        .with_conn(|c| Ok(db::get_setting(c, "gallerydl_bin")))
        .ok()
        .flatten()
        .filter(|s| !s.is_empty());
    override_path.unwrap_or_else(|| {
        config
            .bin_dir()
            .join(gallerydl_filename())
            .to_string_lossy()
            .into()
    })
}

pub fn get_ffmpeg_path(db: &Db, config: &Config) -> String {
    let override_path = db
        .with_conn(|c| Ok(db::get_setting(c, "ffmpeg_bin")))
        .ok()
        .flatten()
        .filter(|s| !s.is_empty());
    override_path.unwrap_or_else(|| {
        config
            .bin_dir()
            .join(ffmpeg_filename())
            .to_string_lossy()
            .into()
    })
}

pub fn get_ffprobe_path(db: &Db, config: &Config) -> String {
    let ffmpeg = get_ffmpeg_path(db, config);
    let p = Path::new(&ffmpeg);
    if let Some(parent) = p.parent() {
        let probe = parent.join(ffprobe_filename());
        if probe.exists() {
            return probe.to_string_lossy().into();
        }
    }
    // Same dir naming convention
    config
        .bin_dir()
        .join(ffprobe_filename())
        .to_string_lossy()
        .into()
}

fn ytdlp_filename() -> &'static str {
    #[cfg(target_os = "windows")]
    {
        "yt-dlp.exe"
    }
    #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
    {
        "yt-dlp_macos"
    }
    #[cfg(all(target_os = "linux", target_arch = "aarch64"))]
    {
        "yt-dlp_linux_aarch64"
    }
    #[cfg(all(not(target_os = "windows"), not(all(target_os = "linux", target_arch = "aarch64")), not(all(target_os = "macos", target_arch = "aarch64"))))]
    {
        "yt-dlp"
    }
}

fn gallerydl_filename() -> &'static str {
    #[cfg(target_os = "windows")]
    {
        "gallery-dl.exe"
    }
    #[cfg(not(target_os = "windows"))]
    {
        "gallery-dl.bin"
    }
}

fn ffmpeg_filename() -> &'static str {
    #[cfg(target_os = "windows")]
    {
        "ffmpeg.exe"
    }
    #[cfg(not(target_os = "windows"))]
    {
        "ffmpeg"
    }
}

fn ffprobe_filename() -> &'static str {
    #[cfg(target_os = "windows")]
    {
        "ffprobe.exe"
    }
    #[cfg(not(target_os = "windows"))]
    {
        "ffprobe"
    }
}

async fn version_of(path: &str, args: &[&str]) -> Option<String> {
    if !Path::new(path).exists() {
        return None;
    }
    let out = tokio::process::Command::new(path)
        .args(args)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .await
        .ok()?;
    let text = String::from_utf8_lossy(&out.stdout);
    let text2 = String::from_utf8_lossy(&out.stderr);
    let combined = if text.trim().is_empty() {
        text2
    } else {
        text
    };
    combined.lines().next().map(|s| s.trim().to_string())
}

pub async fn check_binary(db: &Db, config: &Config, name: &str) -> BinaryStatus {
    let path = match name {
        "ytdlp" => get_ytdlp_path(db, config),
        "gallery-dl" => get_gallerydl_path(db, config),
        "ffmpeg" => get_ffmpeg_path(db, config),
        _ => String::new(),
    };
    let exists = Path::new(&path).exists();
    let version = if exists {
        match name {
            "ytdlp" => version_of(&path, &["--version"]).await,
            "gallery-dl" => version_of(&path, &["--version"]).await,
            "ffmpeg" => version_of(&path, &["-version"]).await,
            _ => None,
        }
    } else {
        None
    };
    BinaryStatus {
        name: name.to_string(),
        path,
        exists,
        version,
    }
}

pub async fn download_binary(
    db: &Db,
    config: &Config,
    name: &str,
) -> anyhow::Result<BinaryStatus> {
    std::fs::create_dir_all(config.bin_dir())?;
    match name {
        "ytdlp" => download_ytdlp(config).await?,
        "gallery-dl" => download_gallerydl(config).await?,
        "ffmpeg" => download_ffmpeg(config).await?,
        _ => anyhow::bail!("unknown binary: {name}"),
    }
    Ok(check_binary(db, config, name).await)
}

async fn download_ytdlp(config: &Config) -> anyhow::Result<()> {
    let asset = ytdlp_filename();
    let url = format!(
        "https://github.com/yt-dlp/yt-dlp/releases/latest/download/{asset}"
    );
    let dest = config.bin_dir().join(asset);
    download_file(&url, &dest).await?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&dest, std::fs::Permissions::from_mode(0o755))?;
    }
    Ok(())
}

async fn download_gallerydl(config: &Config) -> anyhow::Result<()> {
    // Codeberg/Forgejo has no GitHub-style /releases/latest/download/{asset}
    // shortcut — resolve the latest tag via the API, then download by tag.
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::limited(10))
        .build()?;
    let release: serde_json::Value = client
        .get("https://codeberg.org/api/v1/repos/mikf/gallery-dl/releases/latest")
        .send()
        .await?
        .error_for_status()?
        .json()
        .await?;
    let tag = release["tag_name"]
        .as_str()
        .ok_or_else(|| anyhow::anyhow!("gallery-dl release missing tag_name"))?;

    let asset = gallerydl_filename();
    let urls = [
        format!("https://codeberg.org/mikf/gallery-dl/releases/download/{tag}/{asset}"),
        format!("https://codeberg.org/mikf/gallery-dl/releases/download/{tag}/gallery-dl.bin"),
        format!("https://codeberg.org/mikf/gallery-dl/releases/download/{tag}/gallery-dl"),
    ];
    let dest = config.bin_dir().join(asset);
    let mut last_err = None;
    for url in &urls {
        match download_file(url, &dest).await {
            Ok(()) => {
                #[cfg(unix)]
                {
                    use std::os::unix::fs::PermissionsExt;
                    std::fs::set_permissions(&dest, std::fs::Permissions::from_mode(0o755))?;
                }
                return Ok(());
            }
            Err(e) => last_err = Some(e),
        }
    }
    Err(last_err.unwrap_or_else(|| anyhow::anyhow!("gallery-dl download failed")))
}

async fn download_ffmpeg(config: &Config) -> anyhow::Result<()> {
    let (url, is_xz) = ffmpeg_archive();
    let ext = if is_xz { "tar.xz" } else { "zip" };
    let tmp = config.bin_dir().join(format!("ffmpeg-tmp.{ext}"));
    download_file(url, &tmp).await?;
    if is_xz {
        extract_ffmpeg_xz(&tmp, &config.bin_dir())?;
    } else {
        extract_ffmpeg_zip(&tmp, &config.bin_dir())?;
    }
    let _ = std::fs::remove_file(&tmp);
    Ok(())
}

fn ffmpeg_archive() -> (&'static str, bool) {
    #[cfg(all(target_os = "linux", target_arch = "aarch64"))]
    {
        (
            "https://github.com/yt-dlp/FFmpeg-Builds/releases/latest/download/ffmpeg-master-latest-linuxarm64-gpl.tar.xz",
            true,
        )
    }
    #[cfg(all(target_os = "linux", not(target_arch = "aarch64")))]
    {
        (
            "https://github.com/yt-dlp/FFmpeg-Builds/releases/latest/download/ffmpeg-master-latest-linux64-gpl.tar.xz",
            true,
        )
    }
    #[cfg(target_os = "macos")]
    {
        ("https://evermeet.cx/ffmpeg/getrelease/ffmpeg/zip", false)
    }
    #[cfg(target_os = "windows")]
    {
        (
            "https://github.com/yt-dlp/FFmpeg-Builds/releases/latest/download/ffmpeg-master-latest-win64-gpl.zip",
            false,
        )
    }
    #[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows")))]
    {
        (
            "https://github.com/yt-dlp/FFmpeg-Builds/releases/latest/download/ffmpeg-master-latest-linux64-gpl.tar.xz",
            true,
        )
    }
}

fn extract_ffmpeg_xz(archive: &Path, bin_dir: &Path) -> anyhow::Result<()> {
    let file = std::fs::File::open(archive)?;
    let dec = xz2::read::XzDecoder::new(file);
    let mut archive = tar::Archive::new(dec);
    let tmp_dir = bin_dir.join("_ffmpeg_extract");
    let _ = std::fs::remove_dir_all(&tmp_dir);
    std::fs::create_dir_all(&tmp_dir)?;
    archive.unpack(&tmp_dir)?;
    copy_ffmpeg_bins(&tmp_dir, bin_dir)?;
    let _ = std::fs::remove_dir_all(&tmp_dir);
    Ok(())
}

fn extract_ffmpeg_zip(archive: &Path, bin_dir: &Path) -> anyhow::Result<()> {
    let file = std::fs::File::open(archive)?;
    let mut zip = zip::ZipArchive::new(file)?;
    let tmp_dir = bin_dir.join("_ffmpeg_extract");
    let _ = std::fs::remove_dir_all(&tmp_dir);
    std::fs::create_dir_all(&tmp_dir)?;
    zip.extract(&tmp_dir)?;
    copy_ffmpeg_bins(&tmp_dir, bin_dir)?;
    let _ = std::fs::remove_dir_all(&tmp_dir);
    Ok(())
}

fn copy_ffmpeg_bins(tmp_dir: &Path, bin_dir: &Path) -> anyhow::Result<()> {
    for entry in walkdir::WalkDir::new(tmp_dir) {
        let entry = entry?;
        let name = entry.file_name().to_string_lossy();
        if name == "ffmpeg"
            || name == "ffprobe"
            || name == "ffmpeg.exe"
            || name == "ffprobe.exe"
        {
            let dest_name = if name.ends_with(".exe") {
                name.to_string()
            } else {
                name.to_string()
            };
            let dest = bin_dir.join(&dest_name);
            std::fs::copy(entry.path(), &dest)?;
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                std::fs::set_permissions(&dest, std::fs::Permissions::from_mode(0o755))?;
            }
        }
    }
    Ok(())
}

async fn download_file(url: &str, dest: &Path) -> anyhow::Result<()> {
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::limited(10))
        .build()?;
    let resp = client.get(url).send().await?.error_for_status()?;
    let bytes = resp.bytes().await?;
    let partial = dest.with_extension("partial");
    tokio::fs::write(&partial, &bytes).await?;
    tokio::fs::rename(&partial, dest).await?;
    Ok(())
}

pub async fn ensure_binaries(db: &Db, config: &Config) {
    std::fs::create_dir_all(config.bin_dir()).ok();
    for name in ["ytdlp", "gallery-dl", "ffmpeg"] {
        let status = check_binary(db, config, name).await;
        let has_override = match name {
            "ytdlp" => db
                .with_conn(|c| Ok(db::get_setting(c, "ytdlp_bin")))
                .ok()
                .flatten()
                .filter(|s| !s.is_empty())
                .is_some(),
            "gallery-dl" => db
                .with_conn(|c| Ok(db::get_setting(c, "gallerydl_bin")))
                .ok()
                .flatten()
                .filter(|s| !s.is_empty())
                .is_some(),
            "ffmpeg" => db
                .with_conn(|c| Ok(db::get_setting(c, "ffmpeg_bin")))
                .ok()
                .flatten()
                .filter(|s| !s.is_empty())
                .is_some(),
            _ => false,
        };
        if !status.exists && !has_override {
            tracing::info!("downloading missing binary: {name}");
            match download_binary(db, config, name).await {
                Ok(_) => {
                    tracing::info!("downloaded binary: {name}");
                }
                Err(e) => {
                    tracing::warn!("failed to download {name}: {e:#}");
                }
            }
        }
    }
}
